### PR TITLE
makefiles: add targets for experimenting with cargo profiles

### PR DESCRIPTION
**Pull request by brson**
## What have you changed? (mandatory)

As part of #4189, this sets up a way to add a 'third' profile beyond the
cargo-default 'dev' and 'release', as well as experiment generally with cargo
profiles by overlaying cargo config files. It does not itself make any
impact on build times.

Cargo config files contain many options beyond Cargo.toml that can affect build
time, and this setup lets us play with them.

This setup is intended to be temporary until a satisfying set of profiles is
found, at which point the config files and makefile rules will be integrated
into Cargo.toml and the existing makefile rules.

The comments within the makefile and bash script explain in more detail.

This creates three new profiles, contained (for now) in config files in etc/:
dev-nopt, dev-opt, and prod. They are intentionally named distinctly
from the stock cargo profiles to avoid confusion.

I've seeded these profiles with what seem like reasonable initial values. Some differences from the current configuration are: all profiles have debuginfo off by default, with a convenient way to turn debuginfo on via the makefiles; dev profilesuse codegen-units=16 (the default), instead of 4.

I still need to add test and bench profiles to this setup.

__Key result: removing debuginfo from the release profile speeds up the build by 17%.__

Other changes and measurements are hard to interpret yet.

For reviewers, I'm interested in somebody reminding me how to set env-vars per target, whereas the targets I've written set them as part of one log command.

## What are the type of the changes? (mandatory)

Engineering. Doesn't affect any existing code.

## How has this PR been tested? (mandatory)

Manual testing and benchmarking.

## Refer to a related PR or issue link (optional)

- profiles - https://github.com/tikv/tikv/issues/4189
- debuginfo - https://github.com/tikv/tikv/issues/4150

## Benchmark result if necessary (optional)

[Timings for each of the 3 added profiles, along with the existing dev and build profiles](https://gist.github.com/brson/b5f53fa735b35d2a455f7b28d91b21e7)

The results there aren't super encouraging. The best result is that the `dev-opt` profile is significantly faster to build than the release profile. I don't think it is tweaked properly yet though, as evidenced in the following benchmark:

[Bechmarking dev-opt against the current bench profile](https://gist.github.com/brson/5cea8b2717551fd27553da055d8fcc05)

The idea here is that we want the dev-opt and bench profiles to be nearly as fast as the release profile, but consistent enough for benchmarking one commit against another during development.

The results of these benchmarks though vary widely. I suspect part of the reason for that, and maybe for some of our past difficulty interpreting benchmarks, is because neither profile (including today's bench profile) has lto on. Using incremental and codegen-units without any lto results in de-optimizations that are impossible to predict.

# Next steps

- Add bench and test profiles to this
- Add more scripting to compare build-time/benchmarks between two profiles
- Add scripting to compare partial rebuilds
- Experiment with compiling rocksdb as a dylib
- Experiment with compiling every crate as a dylib
- Add sccache support to these scripts
- Profile off-cpu time to see when there are cpus available 
- Profile which processes during the build are taking the longest

Some of these have issues filed, others I will file after getting feedback.

I've also learned more about the benchmarking workflow recently, where the release builders are used to build a release builder, then those builds are sent through some automated benchmarking. There seems to be a requirement that we benchmark fully-optimized builds, which leaves few options for speeding up the build. I think this automated benchmarking workflow is amenable to aggressive caching though since it's mostly rebuilding the same thing all the time. I'll start working on that next.

I'm interested in opinions about what I should pursue next. There are many options.

Fixes https://github.com/tikv/tikv/issues/4189

cc @siddontang @nrc @overvenus @BusyJay

### CODE REVIEW

**Comment by siddontang** [comment]
cool @brson 

can you show us the different time that dev-noopt, dev-opt and prod take?

**Comment by siddontang** [comment]
I think we can try `Add sccache support to these scripts` next at first, it is very simple. We have tried it before but to our surprise, this can't reduce the compiling time too much, very confused. 

Then we can try 

```
Experiment with compiling rocksdb as a dylib
Experiment with compiling every crate as a dylib
```

I guess linking takes a long time too. 

You also told me yesterday that removing debug symbol can reduce 17% compiling time, I think we can use it too. Mostly, in our benchmarking scenarios, we don't need this.

File: etc/cargo.config.dev-opt [hunk 1]
```diff
@@ -0,0 +1,10 @@
+[profile.release]
+opt-level = 2
+debug = false
+rpath = false
+lto = false
+debug-assertions = false
+codegen-units = 16
```
Comment (breeswish) [inline etc/cargo.config.dev-opt]: We met performance regression before when codegen-units is larger than 1

**Comment by brson** [comment]
> can you show us the different time that dev-noopt, dev-opt and prod take?

@siddontang [this link](https://gist.github.com/brson/b5f53fa735b35d2a455f7b28d91b21e7) contains that information.

File: etc/cargo.config.dev-opt (hunk 1 above)
Comment (brson) [reply etc/cargo.config.dev-opt]: Yeah, I've heard that. I'll rebenchmark with the current toolchain before landing anything that affects the real build.

**Comment by brson** [comment]
@overvenus re dynamic linking I found that compiling just rocksdb as a dylib had [insignificant effect[(https://github.com/tikv/tikv/issues/4338#issuecomment-471804746) on full compile times (I did not test partial rebuild times). We've made some [progress](https://github.com/tikv/tikv/issues/4151#issuecomment-473087545) toward dynamically linking tikv.

I suspect that turning just tikv into a dylib won't help build times much since it would just punt all the heavy static linking to building the dylib (this would though eliminate redundant linking for all three bins + tests + benches). If we can though get _every_ crate to build as a dylib that would eliminate a lot of static linking time for tikv-server.

That said, if every library is statically linked then there's going to be no LTO, and the runtime performance will probably be unacceptable.

@overvenus @siddontang I'd like to land this as I now have a fair bit of follow-up results that I'd like to start incorporating.

I think we're getting close to narrowing down a set of config changes that will measurably decrease turnaround for benchmarking without slowing down run-time performance unacceptably.

File: etc/cargo.config.prod [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+# benchmark against a fully-optimized build.
+
+[profile.release]
+opt-level = 3
+debug = false
+rpath = false
+lto = true
+debug-assertions = false
+codegen-units = 1
+panic = 'unwind'
```
Comment (overvenus) [inline etc/cargo.config.prod]: TiKV has its own panic hook (unwind to log), maybe we can just set it to `panic=abort`?

File: etc/cargo.config.prod [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# Configuration for release builds, that is bulids that are
+# _literally_ for packaging and releasing; and cases where we _must_
+# benchmark against a fully-optimized build.
+
+[profile.release]
+opt-level = 3
+debug = false
```
Comment (overvenus) [inline etc/cargo.config.prod]: TiKV may need debug info in order to display specific line when unwinding, or maybe can we generate debuginfo separate?

**Comment by brson** [comment]
> Can we enable LTO for each crate when building statically linked library? Actually we dont crate runtime performance when tikv is built with a debug profile, so we can always disable LTO in debug.

In my comment you are replying to, I think I accidentally said "statically" when I meant "dynamically". That is, "if every crate is linked dynamically there won't be any LTO".

First, yes, we can always disable LTO in debug mode.

Assuming your question then is "can we enable LTO for each crate when building a _dynamically_ linked library", then I think the answer is yes, in one particular scheme.

Before I said

> I suspect that turning just tikv into a dylib won't help build times much since it would just punt all the heavy static linking to building the dylib (this would though eliminate redundant linking for all three bins + tests + benches).

While I was dismissive of this before, I think this may be the _only_ way to get any useful compile time advantage from dylibs without losing LTO. We LTO everything into the tikv dylib, then all the various bins we create don't need an LTO step. This would probably help significantly when running e.g. `cargo bench`, assuming LTO is turned on for the bench profile. Today LTO is surprisingly **off** for the bench profile, which calls any results into question. It might be off specifically because it was so slow to link all the benchmarks.

I think we should turn some form of LTO on for the bench profile, but I'm hopeful thin LTO is viable to keep performance consistent.

> Off topic: another way to reduce build time is to eliminate unnecessary dependency, eg,
> 
>     * Remove tokio-core by replacing it with tokio-threadpool or futures-cpupool,
> 
>     * Remove tokio by replacing hyper 0.12 with hyper 0.9

Yes, this is definitely worth trying. I think though that the results will be small - the biggest crates by far in tikv are rocksdb, kvproto, and tikv.

File: etc/cargo.config.prod (hunk 2 above)
Comment (brson) [reply etc/cargo.config.prod]: I think we might indeed be able to turn on panic=abort in production, though I'd like to look at it closer before doing so. Mostly I want to make sure that the way the abort happens looks intentional from a user perspective and not just like a SIGILL or something else alarming (I'm thinking of e.g. calling process::exit from the panic hook); but also I'd like to verify that there is no scenario where production tikv needs unwinding, including in tikv-ctl and tikv-importer, which are presumably less carefully tested than tikv-server.

We can definitely turn on panic=abort for any build that is just for benchmarking.

File: etc/cargo.config.prod (hunk 3 above)
Comment (brson) [reply etc/cargo.config.prod]: Yeah, this is something I need to look at closer, though I think we are going to have to keep debug=true for release builds, but not for benchmarking builds. I've pushed a change to set debug=true.

We can't generate debuginfo separately.

Per https://github.com/tikv/tikv/issues/4202, where I expect to end up is distributing bins without debuginfo, but such that the debuginfo can be distributed on the side when needed. To do that though we have to have rustc generate the libraries with debuginfo, then strip the debuginfo.

**Comment by brson** [comment]
PTAL @overvenus

**Comment by siddontang** [comment]
can we create a `scripts` directory and move all scripts to it, not in `etc`?

**Comment by brson** [comment]
@siddontang I've moved the scripts into scripts/

Still testing locally.

**Comment by siddontang** [comment]
PTAL @BusyJay @breeswish

**Comment by brson** [comment]
Here are some timings for incremental rebuild:

Gathering rebuild timings for incremental per https://github.com/tikv/tikv/issues/4155#issuecomment-475051840

command: cargo build --release && echo >> src/lib.rs && time cargo build --release
toolchain: rustc 1.35.0-nightly (a9da8fc9c 2019-03-04)
platform: Linux

debug=false, panic=abort, opt-level=3, codegen-units=1, incremental=false, lto="thin":

real    7m43.300s
user    17m22.941s
sys     0m9.721s

debug=false, panic=abort, opt-level=3, codegen-units=1, **incremental=false**, lto="thin":

real    3m34.104s
user    13m10.262s
sys     0m10.384s

So that's very good.

**Comment by brson** [comment]
Closing in favor of https://github.com/tikv/tikv/pull/4415.

**Pull request closed without merging**